x/exchange/keeper/market.go

```diff
@@ -1537,8 +1539,19 @@ func (k Keeper) GetMarketBrief(ctx sdk.Context, marketID uint32) *exchange.Marke
 // WithdrawMarketFunds transfers funds from a market account to another account.
 // The caller is responsible for making sure this withdrawal should be allowed (e.g. by calling CanWithdrawMarketFunds first).
 func (k Keeper) WithdrawMarketFunds(ctx sdk.Context, marketID uint32, toAddr sdk.AccAddress, amount sdk.Coins, withdrawnBy string) error {
+	admin, err := sdk.AccAddressFromBech32(withdrawnBy)
+	if err != nil {
+		return fmt.Errorf("invalid admin %q: %w", withdrawnBy, err)
+	}
+	if k.bankKeeper.BlockedAddr(toAddr) {
+		return fmt.Errorf("%s is not allowed to receive funds", toAddr)
+	}
 	marketAddr := exchange.GetMarketAddress(marketID)
-	err := k.bankKeeper.SendCoins(ctx, marketAddr, toAddr, amount)
+	xferCtx := markertypes.WithTransferAgent(ctx, admin)
+	if toAddr.Equals(admin) {
```

Comment:
This feels like logic that we shouldn't have to understand in the exchange module ... for the sdk50 version we might consider a special ante handler that takes care of this type of setup instead of having to ensure the restrictions are implemented at every use in every module.

Reply:
I don't think we can get away with taking this logic out of the exchange module. At some level, something would have to identify that, for this endpoint (and a couple others), the context should have a transfer agent. There isn't anything that automatically identifies a field as being a transfer agent, it's something that needs to be explicitly identified. Since that identification needs to be somewhere, I feel its best to put it right where that functionality is needed.

The thing I like about having it here is that it's clear that there's a transfer agent for this send, and its only this send that has a transfer agent (in this flow at least). Another thing I like is that I can have unit tests that specifically ensure that a transfer agent is being used. If we rely on an antehandler for this, not only would it be difficult to test that a transfer agent is being used when it should, but we'd also have to add the transfer agent to the context for some unit tests; either way requires understanding of transfer agents in the exchange module.

Ultimately, I feel that, since transfer agents are so rare, it'd be better to explicitly identify them where they're needed, rather than obfuscate their use.